bin/propolis-server/src/lib/vm.rs

```diff
@@ -0,0 +1,1383 @@
+    ///
+    /// This routine fails if the VM was not marked as a migration source or if
+    /// it has another pending request that precludes migration. Note that this
+    /// routine does not fail if the future returned from `upgrade_fn` fails to
+    /// produce a connection to the destination.
+    ///
+    /// On success, clients may query the instance's migration status to
+    /// determine how the migration has progressed.
+    pub fn request_migration_from<F>(
+        &self,
+        migration_id: Uuid,
+        upgrade_fn: F,
+    ) -> Result<(), VmControllerError>
+    where
+        F: FnOnce() -> Result<hyper::upgrade::OnUpgrade, MigrateError>,
+    {
+        let mut inner = self.worker_state.inner.lock().unwrap();
+        match inner.lifecycle_stage {
+            LifecycleStage::NotStarted(_) => {
+                assert!(!inner.marked_migration_source);
+                Err(VmControllerError::InstanceNotActive)
+            }
+            LifecycleStage::Active => {
+                if !inner.marked_migration_source {
+                    Err(VmControllerError::NotMarkedMigrationSource)
+                } else if inner.migrate_from_pending {
+                    Err(VmControllerError::AlreadyMigrationSource)
+                } else if inner.halt_pending {
+                    Err(VmControllerError::InstanceHaltPending)
+                } else {
+                    let upgrade_fut = upgrade_fn()?;
+                    inner.migrate_from_pending = true;
+
+                    // Update the migration state before the task starts so that
+                    // requests to query the migration state will succeed. This
+                    // ensures that if a later request to change the instance's
+                    // state fails because a migration is in progress, there
+                    // will in fact appear to be a pending migration even if the
+                    // state worker hasn't yet picked up the work.
+                    inner.migration_state =
+                        Some((migration_id, ApiMigrationState::Sync));
+                    inner.external_request_queue.push_back(
+                        ExternalRequest::MigrateAsSource {
+                            migration_id,
+                            upgrade_fut,
+                        },
+                    );
+                    self.worker_state.cv.notify_one();
+                    Ok(())
+                }
+            }
+            LifecycleStage::NoLongerActive => {
+                Err(VmControllerError::InstanceNotActive)
+            }
+        }
+    }
+
+    /// Asks to queue a request to start a destination migration task for this
+    /// VM. The migration will have the supplied `migration_id` and will obtain
+    /// its connection to the source by calling `upgrade_fn` to obtain a future
+    /// that yields the necessary connection.
+    ///
+    /// This routine fails if the VM has already been started or if a previous
```

Comment:
"if the VM has already been stared". Does this mean migrations can't happen if a VM is started, or does this mean that the routine will fail if a migration has already started?

Reply:
If a destination VM has already started, then it is invalid as a target for a migration. So I'd say that yes, if a VM has started, then a migration into that VM cannot happen.

Reply:
Ah, so it's "if the destination VM has already been started".  I was not sure which VM we were talking about.

Reply:
In a0b434d I changed the comment around a bit (s/been started/begun to run) to clarify exactly which state transition this routine wants to check for.